README.md
taskwarrior Cookbook
====================

Installs and configures taskwarrior client and server software.

Requirements
------------
All the requirements are installed for you if you use [Berkshelf](http://berkshelf.com/). Altough it is
good practice to add the following to the role run list.

#### cookbooks
- `apt` - To install the lastest packages.
- `git` - To clone the code both for client and server source installs.
- `cmake` - Build dependency for task and taskserver.
- `build-essential` - Build dependency for task and taskserver.
- `perl` - Build dependency for taskserver.
- `python` - Build dependency for taskserver.
- `runit` - To manage the taskserver daemon.

#### Plaforms
- Ubuntu
- Debian

Attributes
----------

#### taskwarrior::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['taskwarrior']['method']</tt></td>
    <td>String</td>
    <td>Choose to install from the repository of from the lastest on the git repository</td>
    <td><tt>package</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["source"]["git_repository"]</tt></td>
    <td>String</td>
    <td>The git repository to clone taskwarrior from.</td>
    <td><tt>git://tasktools.org/task.git</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["source"]["git_revision"]</tt></td>
    <td>String</td>
    <td>The git revision to build from, HEAD is lastest</td>
    <td><tt>HEAD</tt></td>
  </tr>
</table>

#### taskwarrior::server
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["git_repository"]</tt></td>
    <td>String</td>
    <td>The git repository to clone taskserver from.</td>
    <td><tt>git://tasktools.org/taskd.git</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["git_revision"]</tt></td>
    <td>String</td>
    <td>The git revision to build from, HEAD is lastest</td>
    <td><tt>HEAD</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["home"]</tt></td>
    <td>String</td>
    <td>The main directory for taskd</td>
    <td><tt>/var/lib/taskd</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["data_dir"]</tt></td>
    <td>String</td>
    <td>Taskd data directory</td>
    <td><tt>default["taskwarrior"]["server"]["home"]}/data</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["keys_dir"]</tt></td>
    <td>String</td>
    <td>Fully qualified path for the keys</td>
    <td><tt>default["taskwarrior"]["server"]["home"]}/keys</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["confirmation"]</tt></td>
    <td>String</td>
    <td>Determines whether certain commands are confirmed</td>
    <td><tt>on</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["extensions"]</tt></td>
    <td>String</td>
    <td>Fully qualified path of the taskd extension scripts. Currently there are none.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["ip_log"]</tt></td>
    <td>String</td>
    <td>Logs the IP addresses of incoming requests.</td>
    <td><tt>on</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["log_dir"]</tt></td>
    <td>String</td>
    <td>Log directory for taskd.</td>
    <td><tt>/var/log/taskd</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["log_file"]</tt></td>
    <td>String</td>
    <td>Main log file.</td>
    <td><tt>default["taskwarrior"]["server"]["log_dir"]}/keys</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["queue_size"]</tt></td>
    <td>Integer</td>
    <td>Size of the connection backlog.</td>
    <td><tt>10</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["request_limit"]</tt></td>
    <td>Integer</td>
    <td>Size limit of incoming requests, in bytes.</td>
    <td><tt>1048576</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["link"]</tt></td>
    <td>String</td>
    <td>The address of the taskd server followed by a colon and the por number.</td>
    <td><tt>localhost:6544</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["organization"]</tt></td>
    <td>String</td>
    <td>The name of your organization to generate the certificates.</td>
    <td><tt>Empty</tt></td>
  </tr>
  <tr>
    <td><tt>["taskwarrior"]["server"]["clients"]</tt></td>
    <td>String</td>
    <td>Regex string to allow client to connect to the server. Check the taskd log for more examples.</td>
    <td><tt>^task [2-9]</tt></td>
  </tr>
</table>

Usage
-----

#### taskwarrior::default

Installs the taskwarrior client that can work standalone, just add it to your run_list. You can choose to install from the repositories,
or to build from source. Just set the ["taskwarrior"]["install_method"] to "source", like so.

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[taskwarrior]"
  ]
  "override_attrubutes": {
    "taskwarrior": {
      "install_method": "source"
    }
  }
}
```

#### taskwarrior::server

Builds and installs the taskwarrior service daemon (taskd) and sets up supervising with runit.
Check the taskwarrior wiki pages for [operation](http://taskwarrior.org/projects/taskwarrior/wiki/Taskserver_Operation), [setup](http://taskwarrior.org/projects/taskwarrior/wiki/Server_setup) amd [ciphers](http://taskwarrior.org/projects/taskwarrior/wiki/Ciphers)

Add the following to you users data bag.
```json
{
  "id": "bob",
  "taskwarrior": {
    "organization": "Public",
    "groups": ["groupA", "groupB"]
  }
}
```

This will generate client keys and the taskd user and its Uuid for bob. The key distrubution at the moment is done by hand so you
need to get them from the key directory.

Contributing
------------

The testing is done mostly with [test-kitchen](http://kitchen.ci/) before everything, set it up. Also [foodcritic](http://www.foodcritic.io/) is used for linting.

- Fork the repository on Github
- Create a named feature branch (like `add_component_x`)
- Write your change
- Make it pass foodcritic.
- Write tests for your change (if applicable)
- Run the tests, ensuring they all pass
- Submit a Pull Request using Github

License and Authors
-------------------
Authors:
 - Alfredo Palhares (masterkorp@masterkorp.net) and masterkorp on irc.freenode.org


CHANGELOG.md
taskwarrior CHANGELOG
=====================

1.1.3
-----
- Fixed missing CA on taskd ([@masterkorp])
- Do not generate certs for users that do not have the taskwarrior attributes
- Allow taskwarrior clients above version 2 to connect

1.1.2
-----
- Fixed runit initial starts because of the log ([@masterkorp][])
- Allow taskserver to write his own logs
- Generate the certificates as non root
- Run taskserver under its on user

0.1.1
-----
- Initial release of taskwarrior ([@masterkorp][])

- - -

<!--- The following link definition list is generated by PimpMyChangelog --->
[@masterkorp]: https://github.com/masterkorp


recipes/certs.rb
package "gnutls-bin"

log "The organization attribute is not set, certificates can't be set." do
  level :fatal
  only_if { node["taskwarrior"]["server"].attribute?("organization") }
end

directory node["taskwarrior"]["server"]["keys_dir"] do
  owner "taskd"
  group "taskd"
  mode 0700
end

bash "Generating CA key" do
  user "taskd"
  cwd node["taskwarrior"]["server"]["keys_dir"]
  code <<-EOH
    certtool --generate-privkey --outfile ca.key.pem
  EOH
  not_if {::File.exists?("#{node["taskwarrior"]["server"]["keys_dir"]}/ca.key.pem")}
end

template "#{node["taskwarrior"]["server"]["keys_dir"]}/ca.info" do
  source "ca.info.erb"
  owner "taskd"
  group "taskd"
  mode 00600
  variables({
    :organization => node["taskwarrior"]["server"]["organization"]
  })
  not_if {::File.exists?("#{node["taskwarrior"]["server"]["keys_dir"]}/ca.cert.pem")}
  notifies :run, "bash[Generating CA Cert]", :immediately
end

bash "Generating CA Cert" do
  user "taskd"
  cwd node["taskwarrior"]["server"]["keys_dir"]
  code <<-EOH
    certtool --generate-self-signed \
    --load-privkey ca.key.pem \
    --template ca.info \
    --outfile ca.cert.pem
    rm ca.info
    [ -s ca.cert.pem ]
  EOH
  action :nothing
end

bash "Generating Server key" do
  user "taskd"
  cwd node["taskwarrior"]["server"]["keys_dir"]
  code <<-EOH
    certtool --generate-privkey --outfile server.key.pem
  EOH
  not_if {::File.exists?("#{node["taskwarrior"]["server"]["keys_dir"]}/server.key.pem")}
end

template "#{node["taskwarrior"]["server"]["keys_dir"]}/server.info" do
  source "server.info.erb"
  owner "taskd"
  group "taskd"
  mode 00600
  variables({
    :organization => node["taskwarrior"]["server"]["organization"],
    :cn => node["ipaddress"]
  })
  not_if {::File.exists?("#{node["taskwarrior"]["server"]["keys_dir"]}/server.cert.pem")}
  notifies :run, "bash[Generating Server Cert]", :immediately
end

bash "Generating Server Cert" do
  user "taskd"
  cwd node["taskwarrior"]["server"]["keys_dir"]
  code <<-EOH
    certtool --generate-certificate \
    --load-privkey server.key.pem \
    --load-ca-certificate ca.cert.pem \
    --load-ca-privkey ca.key.pem \
    --template server.info \
    --outfile server.cert.pem
    rm server.info
    [ -s server.cert.pem ]
  EOH
  action :nothing
end

template "#{node["taskwarrior"]["server"]["keys_dir"]}/crl.info" do
  source "crl.info.erb"
  owner "taskd"
  group "taskd"
  mode 00600
  variables({
    :expiration_days => "365"
  })
  not_if {::File.exists?("#{node["taskwarrior"]["server"]["keys_dir"]}/server.crl.pem")}
  notifies :run, "bash[Generating Server CRL]", :immediately
end

bash "Generating Server CRL" do
  user "taskd"
  cwd node["taskwarrior"]["server"]["keys_dir"]
  code <<-EOH
    certtool --generate-crl \
    --load-ca-privkey ca.key.pem \
    --load-ca-certificate ca.cert.pem \
    --template crl.info \
    --outfile server.crl.pem
    rm crl.info
    [ -s server.crl.pem ]
  EOH
  action :nothing
end
users = data_bag("users")

users.each() do |s|
  u = data_bag_item("users", s)

  if not u["taskwarrior"].nil?
    bash "Create Organization for the user" do
      user "taskd"
      cwd node["taskwarrior"]["server"]["home"]
      code <<-EOH
        taskd add org #{u["taskwarrior"]["organization"]} --data #{node["taskwarrior"]["server"]["data_dir"]}
      EOH
      not_if do ::File.directory?("#{node["taskwarrior"]["server"]["data_dir"]}/orgs/#{u["taskwarrior"]["organization"]}") end
    end

    directory "#{node["taskwarrior"]["server"]["keys_dir"]}/#{u["id"]}" do
      owner "taskd"
      group "taskd"
      notifies :run, "bash[Create user and key]", :immediately
    end

    bash "Create user and key" do
      user "taskd"
      cwd "#{node["taskwarrior"]["server"]["keys_dir"]}/#{u["id"]}"
      code <<-EOH
        taskd add user #{u["taskwarrior"]["organization"]} #{u["id"]} \
        --data #{node["taskwarrior"]["server"]["data_dir"]} \
        2>&1 | tee #{u["id"]}.txt
        certtool --generate-privkey --outfile #{u["id"]}.key.pem
      EOH
      action :nothing
    end

    template "#{node["taskwarrior"]["server"]["keys_dir"]}/#{u["id"]}/client.info" do
      source "client.info.erb"
      owner "taskd"
      group "taskd"
      mode 00600
      variables({
        :organization => node["taskwarrior"]["server"]["organization"],
        :cn => node["ipaddress"]
      })
      not_if {::File.exists?("#{node["taskwarrior"]["server"]["keys_dir"]}/#{u["id"]}/#{u["id"]}.cert.pem")}
      notifies :run, "bash[Generating User Cert]", :immediately
    end

    bash "Generating User Cert" do
      user "taskd"
      cwd "#{node["taskwarrior"]["server"]["keys_dir"]}/#{u["id"]}"
      code <<-EOH
        certtool --generate-certificate \
        --load-privkey #{u["id"]}.key.pem \
        --load-ca-certificate ../ca.cert.pem \
        --load-ca-privkey ../ca.key.pem \
        --template client.info \
        --outfile #{u["id"]}.cert.pem
        rm client.info
        [ -s #{u["id"]}.cert.pem ]
      EOH
      action :nothing
    end
  else
    log "The #{u["id"]} does not have an Taskwarrior attribute so it will not be created" do
      level :warn
    end
  end
end




recipes/server.rb
include_recipe "apt"
include_recipe "git"
include_recipe "build-essential"
include_recipe "cmake"
include_recipe "perl"
include_recipe "python"
include_recipe "runit"

%w{ libgnutls-dev libreadline6 uuid-dev }.each do |pkg|
  package pkg
end

git "#{Chef::Config[:file_cache_path]}/taskd.git" do
  repository node["taskwarrior"]["server"]["git_repository"]
  reference node["taskwarrior"]["server"]["git_revision"]
  action :sync
  notifies :run, "bash[Install taskd]", :immediately
end

bash "Install taskd" do
  user "root"
  cwd "#{Chef::Config[:file_cache_path]}/taskd.git"
  code <<-EOH
  cmake .
  make
  make install
  EOH
  action :nothing
  notifies :restart, "runit_service[taskd]", :delayed
end

user "taskd" do
  system true
  home node["taskwarrior"]["server"]["home"]
  shell "/bin/bash"
end

group "log" do
  members "taskd"
  append true
end

directory node["taskwarrior"]["server"]["home"] do
  owner "taskd"
  group "taskd"
  mode 00744
  recursive true
end

directory node["taskwarrior"]["server"]["data_dir"] do
  owner "taskd"
  group "taskd"
  mode 00744
  recursive true
end

directory node["taskwarrior"]["server"]["log_dir"] do
  owner "taskd"
  group "taskd"
  mode 00744
  recursive true
end

bash "Initialize database" do
  user "taskd"
  cwd node["taskwarrior"]["server"]["home"]
  code <<-EOH
  taskd init --data #{node["taskwarrior"]["server"]["data_dir"]}
  EOH
  not_if { ::File.exists?("#{node["taskwarrior"]["server"]["data_dir"]}/org") }
end

include_recipe "taskwarrior::certs"

template "#{node["taskwarrior"]["server"]["data_dir"]}/config" do
  source "taskd.config.erb"
  owner "taskd"
  group "taskd"
  mode 00644
  variables({
    :confirmation => node["taskwarrior"]["server"]["confirmation"],
    :extensions => node["taskwarrior"]["server"]["extensions"],
    :ip_log => node["taskwarrior"]["server"]["ip_log"],
    :log => node["taskwarrior"]["server"]["log_file"],
    :queue_size=> node["taskwarrior"]["server"]["queue_size"],
    :request_limit=> node["taskwarrior"]["server"]["request_limit"],
    :root => node["taskwarrior"]["server"]["data_dir"],
    :server => node["taskwarrior"]["server"]["link"],
    :server_cert => "#{node["taskwarrior"]["server"]["keys_dir"]}/server.cert.pem",
    :server_crl => "#{node["taskwarrior"]["server"]["keys_dir"]}/server.crl.pem",
    :server_key => "#{node["taskwarrior"]["server"]["keys_dir"]}/server.key.pem",
    :ca_cert => "#{node["taskwarrior"]["server"]["keys_dir"]}/ca.cert.pem",
    :client_allow => node["taskwarrior"]["server"]["clients"]
  })
  notifies :restart, "runit_service[taskd]", :delayed
end


runit_service "taskd" do
  options({
    :user => "taskd",
    :data_dir => node["taskwarrior"]["server"]["data_dir"]
  })
  log false
end


attributes/default.rb


default["taskwarrior"]["source"]["git_repository"] = "git://tasktools.org/task.git"
default["taskwarrior"]["source"]["git_revision"] = "HEAD"

default["taskwarrior"]["server"]["git_repository"] = "git://tasktools.org/taskd.git"
default["taskwarrior"]["server"]["git_revision"] = "HEAD"
default["taskwarrior"]["server"]["home"] = "/var/lib/taskd"
default["taskwarrior"]["server"]["data_dir"] = "#{default["taskwarrior"]["server"]["home"]}/data"
default["taskwarrior"]["server"]["keys_dir"] = "#{default["taskwarrior"]["server"]["home"]}/keys"

default["taskwarrior"]["server"]["confirmation"] = "on"
default["taskwarrior"]["server"]["extensions"] = ""
default["taskwarrior"]["server"]["ip_log"] = "on"
default["taskwarrior"]["server"]["log_dir"] = "/var/log/taskd"
default["taskwarrior"]["server"]["log_file"] = "#{default["taskwarrior"]["server"]["log_dir"]}/taskd.log"
default["taskwarrior"]["server"]["queue_size"] = 10
default["taskwarrior"]["server"]["request_limit"] = 1048576
default["taskwarrior"]["server"]["link"] = "localhost:6544"
default["taskwarrior"]["server"]["clients"] = "^task [2-9]"
